osu.Server.DifficultyCalculator/Program.cs

```diff
@@ -0,0 +1,214 @@
+
+        private Database database;
+
+
+        public void OnExecute(CommandLineApplication app, IConsole console)
+        {
+            database = new Database(connection_string);
+
+            using (var reader = database.RunQuery("SELECT attrib_id, name FROM osu_difficulty_attribs"))
+            {
+                while (reader.Read())
+                    attributeIds[reader.GetString(1)] = reader.GetInt32(0);
+            }
+
+            var tasks = new List<Task>();
+
+            using (var reader = database.RunQuery("SELECT osu_beatmaps.beatmap_id, checksum FROM osu_beatmaps ORDER BY beatmap_id ASC"))
+            {
+                while (reader.Read())
+                {
+                    int id = reader.GetInt32(0);
+                    string hash = reader.IsDBNull(1) ? string.Empty : reader.GetString(1);
+                    tasks.Add(processBeatmap(id));
+                }
+            }
+
+            Task.WaitAll(tasks.ToArray());
+        }
+
+        private async Task processBeatmap(int beatmapId)
+        {
+            string path = Path.Combine(beatmaps_path, beatmapId + ".osu");
+            if (!File.Exists(path))
+                return;
+
+            await Task.Run(async () =>
+            {
+                var localBeatmap = new LocalWorkingBeatmap(path);
+
+                // Todo: For each ruleset
+                var playable = localBeatmap.GetPlayableBeatmap(localBeatmap.BeatmapInfo.Ruleset);
+                var ruleset = localBeatmap.BeatmapInfo.Ruleset.CreateInstance();
+
+                foreach (var mod in ruleset.GetModsFor(ModType.DifficultyCalculation))
+                {
+                    await Task.Run(() =>
+                    {
+                        var legacyMod = toLegacyMod(mod);
+
+                        var attributes = new Dictionary<string, object>();
+                        double starRating = ruleset.CreateDifficultyCalculator(playable, toModArray(mod)).Calculate(attributes);
+
+                        database.RunNonQuery(
+                            "INSERT INTO osu_beatmap_difficulty (beatmap_id, mode, mods, diff_unified) VALUES (@beatmap_id, @mode, @mods, @diff_unified) ON DUPLICATE KEY UPDATE diff_unified = @diff_unified",
+                            new MySqlParameter("@beatmap_id", beatmapId),
+                            new MySqlParameter("@mode", ruleset.RulesetInfo.ID),
+                            new MySqlParameter("@mods", (int)legacyMod),
+                            new MySqlParameter("@diff_unified", starRating));
+
+                        if (attributes.Count > 0)
+                        {
+                            string command = "INSERT INTO osu_beatmap_difficulty_attribs (beatmap_id, mode, mods, attrib_id, value) VALUES ";
+
+                            var parameters = new List<MySqlParameter>
+                            {
+                                new MySqlParameter("@beatmap_id", beatmapId),
```

Comment:
The newer helper methods in Mysql.Data don't require explicit `MySqlParameter` construction for each parameter.

Reply:
Which helper methods are you referring to, exactly?

Reply:
Check out @notbakaneko's PR for some examples (https://github.com/ppy/osu-server/pull/1/files#diff-699987b80d825881c4813b3be5f28a21R35).